concepts/materialized_views.md

```diff
@@ -0,0 +1,378 @@
+
+There are two approaches with regard to timing of view maintenance: immediate and deferred. In immediate maintenance,
+views are updated in the same transaction that its base table is modified.  In deferred maintenance, views are updated
+after the transaction is committed, for example, when the view is accessed, as a response to user command like
+`REFRESH MATERIALIZED VIEW`, or periodically in background, and so on. `pg_ivm` provides a kind of immediate
+maintenance, in which materialized views are updated immediately in AFTER triggers when a base table is modified.
+
+We call a materialized view supporting IVM an Incrementally Maintainable Materialized View (IMMV). To create IMMV,
+you have to call `create_immv` function with a relation name and a view definition query. For example:
+
+```sql
+SELECT create_immv('sales_test', 'SELECT * FROM sales');
+```
+
+creates an IMMV with name 'sales_test' defined as `SELECT * FROM sales`. This corresponds to the following command
+to create a normal materialized view:
+
+```sql
+CREATE MATERIALIZED VIEW sales_test AS SELECT * FROM sales;
+```
+
+When an IMMV is created, some triggers are automatically created so that the view's contents are immediately updated
+when its base tables are modified.
+
+```sql
+postgres=# SELECT create_immv('m', 'SELECT * FROM t0');
+NOTICE:  could not create an index on immv "m" automatically
+DETAIL:  This target list does not have all the primary key columns, or this view does not contain DISTINCT clause.
+HINT:  Create an index on the immv for efficient incremental maintenance.
+ create_immv
+-------------
+           3
+(1 row)
+
+postgres=# SELECT * FROM m;
+ i
+---
+ 1
+ 2
+ 3
+(3 rows)
+
+postgres=# INSERT INTO t0 VALUES (4);
+INSERT 0 1
+postgres=# SELECT * FROM m; -- automatically updated
+ i
+---
+ 1
+ 2
+ 3
+ 4
+(4 rows)
+```
+### Functions
+
+#### create_imm
+
+Use `create_immv` function to create IMMV.
+```
+create_immv(immv_name text, view_definition text) RETURNS bigint
+```
+`create_immv` defines a new IMMV of a query. A table of the name `immv_name` is created and a query specified by
+`view_definition` is executed and used to populate the IMMV. The query is stored in `pg_ivm_immv`, so that it can be
+refreshed later upon incremental view maintenance. `create_immv` returns the number of rows in the created IMMV.
+
+When an IMMV is created, triggers are automatically created so that the view's contents are immediately updated when
+its base tables are modified. In addition, a unique index is created on the IMMV automatically if possible.  If the
+view definition query has a GROUP BY clause, a unique index is created on the columns of GROUP BY expressions. Also,
+if the view has DISTINCT clause, a unique index is created on all columns in the target list. Otherwise, if the IMMV
+contains all primary key attributes of its base tables in the target list, a unique index is created on these
+attributes. In other cases, no index is created.
+
+#### refresh_imm
```

Comment:
same as the create_imm - having the section name be different than the function is confusing